File: src/java/org/orbeon/oxf/transformer/xupdate/statement/Variable.java
/**
 *  Copyright (C) 2004 Orbeon, Inc.
 *
 *  This program is free software; you can redistribute it and/or modify it under the terms of the
 *  GNU Lesser General Public License as published by the Free Software Foundation; either version
 *  2.1 of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *  See the GNU Lesser General Public License for more details.
 *
 *  The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.transformer.xupdate.statement;

import org.dom4j.QName;
import org.jaxen.NamespaceContext;
import org.orbeon.oxf.transformer.xupdate.Statement;
import org.orbeon.oxf.transformer.xupdate.VariableContextImpl;
import org.orbeon.oxf.transformer.xupdate.DocumentContext;
import org.orbeon.oxf.xml.dom4j.LocationData;

import javax.xml.transform.URIResolver;

public class Variable extends Statement {

    private QName qname;
    private String select;
    private NamespaceContext namespaceContext;
    private Statement[] statements;

    public Variable(LocationData locationData, QName qname, String select, NamespaceContext namespaceContext, Statement[] statements) {
        super(locationData);
        this.qname = qname;
        this.select = select;
        this.namespaceContext = namespaceContext;
        this.statements = statements;
    }

    public Object execute(URIResolver uriResolver, Object context, VariableContextImpl variableContext, DocumentContext documentContext) {
        if (select != null) {
            return Utils.evaluate(uriResolver, context, variableContext, documentContext, getLocationData(), select, namespaceContext);
        } else {
            return Utils.execute(uriResolver, context, variableContext, documentContext, statements);
        }
    }

    public QName getName() {
        return qname;
    }
}

File: xforms/jvm/src/main/scala/org/orbeon/oxf/xforms/Variable.scala
/**
  * Copyright (C) 2010 Orbeon, Inc.
  *
  * This program is free software; you can redistribute it and/or modify it under the terms of the
  * GNU Lesser General Public License as published by the Free Software Foundation; either version
  * 2.1 of the License, or (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
  * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  * See the GNU Lesser General Public License for more details.
  *
  * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
  */
package org.orbeon.oxf.xforms

import org.orbeon.datatypes.LocationData
import org.orbeon.oxf.common.OXFException
import org.orbeon.oxf.util.XPathCache
import org.orbeon.oxf.xforms.analysis.ElementAnalysis
import org.orbeon.oxf.xforms.analysis.controls.{VariableAnalysis, VariableAnalysisTrait}
import org.orbeon.saxon.om.ValueRepresentation
import org.orbeon.saxon.value.{EmptySequence, StringValue}
import org.orbeon.xforms.XFormsId

import scala.util.control.NonFatal

// Represent an xf:var.
class Variable(val staticVariable: VariableAnalysisTrait, val containingDocument: XFormsContainingDocument) {

  private var variableValueOpt: Option[ValueRepresentation] = None

  def valueEvaluateIfNeeded(
    contextStack      : XFormsContextStack,
    sourceEffectiveId : String,
    pushOuterContext  : Boolean,
    handleNonFatal    : Boolean
  ): ValueRepresentation =
    variableValueOpt getOrElse {
      val result =
        evaluate(
          contextStack,
          XFormsId.getRelatedEffectiveId(sourceEffectiveId, staticVariable.valueStaticId),
          pushOuterContext,
          handleNonFatal
        )
      variableValueOpt = Some(result)
      result
    }

  def markDirty(): Unit             = variableValueOpt = None
  def getLocationData: LocationData = staticVariable.asInstanceOf[ElementAnalysis].locationData

  private def evaluate(
    contextStack      : XFormsContextStack,
    sourceEffectiveId : String,
    pushOuterContext  : Boolean,
    handleNonFatal    : Boolean
  ): ValueRepresentation =
    staticVariable.expressionStringOpt match {
      case None =>
        // Inline constructor (for now, only textual content, but in the future, we could allow xf:output in it? more?)
        new StringValue(staticVariable.valueElement.getStringValue)
      case Some(expression) =>

        // Push binding for evaluation, so that @context and @model are evaluated
        val pushContext = pushOuterContext || staticVariable.hasNestedValue
        if (pushContext)
          contextStack.pushBinding(staticVariable.valueElement, sourceEffectiveId, staticVariable.valueScope)

        val result = {
          val bindingContext = contextStack.getCurrentBindingContext
          val currentNodeset = bindingContext.nodeset
          if (! currentNodeset.isEmpty) {
            // TODO: in the future, we should allow null context for expressions that do not depend on the context
            val functionContext = contextStack.getFunctionContext(sourceEffectiveId)
            val scopeModelVariables = VariableAnalysis.variableScopesModelVariables(staticVariable)
            try {
              XPathCache.evaluateAsExtent(
                currentNodeset,
                bindingContext.position,
                expression,
                staticVariable.valueNamespaceMapping,
                bindingContext.getInScopeVariables(scopeModelVariables),
                containingDocument.functionLibrary,
                functionContext,
                null,
                getLocationData,
                containingDocument.getRequestStats.getReporter
              )
            } catch {
              case NonFatal(t) if handleNonFatal =>
                  // Don't consider this as fatal
                  // Default value is the empty sequence
                  XFormsError.handleNonFatalXPathError(contextStack.container, t)
                  EmptySequence.getInstance
              case NonFatal(t) =>
                throw new OXFException(t)
            }
          } else {
            EmptySequence.getInstance
          }
        }

        if (pushContext)
          contextStack.popBinding

        result
    }
}
